qa/tasks/daemonwatchdog.py

```diff
@@ -62,11 +75,11 @@ def bark(self):
                 except:
                     self.logger.exception("ignoring exception:")
         daemons = []
-        daemons.extend(filter(lambda daemon: daemon.running() and not daemon.proc.finished, self.ctx.daemons.iter_daemons_of_role('osd', cluster=self.cluster)))
-        daemons.extend(filter(lambda daemon: daemon.running() and not daemon.proc.finished, self.ctx.daemons.iter_daemons_of_role('mds', cluster=self.cluster)))
-        daemons.extend(filter(lambda daemon: daemon.running() and not daemon.proc.finished, self.ctx.daemons.iter_daemons_of_role('mon', cluster=self.cluster)))
-        daemons.extend(filter(lambda daemon: daemon.running() and not daemon.proc.finished, self.ctx.daemons.iter_daemons_of_role('rgw', cluster=self.cluster)))
-        daemons.extend(filter(lambda daemon: daemon.running() and not daemon.proc.finished, self.ctx.daemons.iter_daemons_of_role('mgr', cluster=self.cluster)))
+        daemons.extend(filter(self.is_running, self.ctx.daemons.iter_daemons_of_role('osd', cluster=self.cluster)))
```

Comment:
huh? Why?

Reply:
This is because when daemons are registered from qa/tasks/ceph.py, they are of type `DaemonState` (see [initialisation without `use_cephadm`](https://github.com/ceph/ceph/blob/4d5b2cdc1652f11ae8427a46b6b5f82112c34208/qa/tasks/ceph.py#L1919)) and when they are registered from `qa/tasks/cephadm.py`, they are of type `CephadmUnit` ([initialised with `use_cephadm`](https://github.com/ceph/ceph/blob/4d5b2cdc1652f11ae8427a46b6b5f82112c34208/qa/tasks/cephadm.py#L1861)). See `DaemonGroup` logic [here](https://github.com/ceph/teuthology/blob/544fecbcd55f3d2b6f478823823ce40cbefef1d4/teuthology/orchestra/daemon/group.py#L63).

The reason why I had to make these changes above is because of difference between implementation of [CephadmUnit](https://github.com/ceph/teuthology/blob/544fecbcd55f3d2b6f478823823ce40cbefef1d4/teuthology/orchestra/daemon/cephadmunit.py#L109) and [DaemonState](https://github.com/ceph/teuthology/blob/544fecbcd55f3d2b6f478823823ce40cbefef1d4/teuthology/orchestra/daemon/state.py#L93) - basically, `CephadmUnit` doesn't have a `proc` object. And so I only want to check `not daemon.proc.finished` only for `DaemonState` daemons. That's why I extracted the logic to a `is_running` function.

Reply:
This type of type-specific behavior (`is_running`/`is_stopped` above) is a code smell. Let's fix teuthology so we can call a single `DaemonState` method for both cases. Note `DaemonState.check_status` would be suitable if were implemented by `CephadmUnit`:

https://github.com/ceph/teuthology/blob/544fecbcd55f3d2b6f478823823ce40cbefef1d4/teuthology/orchestra/daemon/state.py#L34

but `DaemonState.check_status` should also be modified:

```
return self.proc.poll() if self.proc is not None else False
```

Reply:
Actually `check_status` may raise an exception. Let's add a method:

```
#DaemonState; override in CephadmUnit
def finished(self):
  return self.proc.finished if self.proc is not None else False
```

Reply:
Done! https://github.com/ceph/teuthology/pull/1961